File: src/core/services/session-bindings.ts
import type { Keybinding, ResolvedConfig } from '../../types';
import type { ConfiguredShortcuts } from '../utils/shortcut-config';
import type {
  CompiledMpvCommandBinding,
  CompiledSessionActionBinding,
  CompiledSessionBinding,
  PluginSessionBinding,
  PluginSessionBindingsArtifact,
  SessionActionId,
  SessionBindingWarning,
  SessionKeyModifier,
  SessionKeySpec,
} from '../../types/session-bindings';
import { SPECIAL_COMMANDS } from '../../config';

type PlatformKeyModel = 'darwin' | 'win32' | 'linux';

type CompileSessionBindingsInput = {
  keybindings: Keybinding[];
  shortcuts: ConfiguredShortcuts;
  statsToggleKey?: string | null;
  statsMarkWatchedKey?: string | null;
  platform: PlatformKeyModel;
  rawConfig?: ResolvedConfig | null;
};

type DraftBinding = {
  binding: CompiledSessionBinding;
  actionFingerprint: string;
};

const MODIFIER_ORDER: SessionKeyModifier[] = ['ctrl', 'alt', 'shift', 'meta'];
const MPV_MOUSE_BUTTON_CODES = new Set([
  'MBTN_LEFT',
  'MBTN_MID',
  'MBTN_RIGHT',
  'MBTN_BACK',
  'MBTN_FORWARD',
]);

const SESSION_SHORTCUT_ACTIONS: Array<{
  key: keyof Omit<ConfiguredShortcuts, 'multiCopyTimeoutMs'>;
  actionId: SessionActionId;
}> = [
  { key: 'toggleVisibleOverlayGlobal', actionId: 'toggleVisibleOverlay' },
  { key: 'copySubtitle', actionId: 'copySubtitle' },
  { key: 'copySubtitleMultiple', actionId: 'copySubtitleMultiple' },
  { key: 'updateLastCardFromClipboard', actionId: 'updateLastCardFromClipboard' },
  { key: 'triggerFieldGrouping', actionId: 'triggerFieldGrouping' },
  { key: 'triggerSubsync', actionId: 'triggerSubsync' },
  { key: 'mineSentence', actionId: 'mineSentence' },
  { key: 'mineSentenceMultiple', actionId: 'mineSentenceMultiple' },
  { key: 'toggleSecondarySub', actionId: 'toggleSecondarySub' },
  { key: 'markAudioCard', actionId: 'markAudioCard' },
  { key: 'openCharacterDictionaryManager', actionId: 'openCharacterDictionaryManager' },
  { key: 'openRuntimeOptions', actionId: 'openRuntimeOptions' },
  { key: 'openJimaku', actionId: 'openJimaku' },
  { key: 'openTsukihime', actionId: 'openTsukihime' },
  { key: 'openSessionHelp', actionId: 'openSessionHelp' },
  { key: 'openControllerSelect', actionId: 'openControllerSelect' },
  { key: 'openControllerDebug', actionId: 'openControllerDebug' },
  { key: 'toggleSubtitleSidebar', actionId: 'toggleSubtitleSidebar' },
  { key: 'toggleNotificationHistory', actionId: 'toggleNotificationHistory' },
  { key: 'appendClipboardVideoToQueue', actionId: 'appendClipboardVideoToQueue' },
];

function normalizeModifiers(modifiers: SessionKeyModifier[]): SessionKeyModifier[] {
  return [...new Set(modifiers)].sort(
    (left, right) => MODIFIER_ORDER.indexOf(left) - MODIFIER_ORDER.indexOf(right),
  );
}

function isValidCommandEntry(value: unknown): value is string | number {
  return typeof value === 'string' || typeof value === 'number';
}

function normalizeCodeToken(
  token: string,
  options: { allowMouseButtons?: boolean } = {},
): string | null {
  const normalized = token.trim();
  if (!normalized) return null;
  if (options.allowMouseButtons === true) {
    const normalizedMouse = normalized.toUpperCase();
    if (MPV_MOUSE_BUTTON_CODES.has(normalizedMouse)) {
      return normalizedMouse;
    }
  }
  if (/^[a-z]$/i.test(normalized)) {
    return `Key${normalized.toUpperCase()}`;
  }
  if (/^[0-9]$/.test(normalized)) {
    return `Digit${normalized}`;
  }

  const exactMap: Record<string, string> = {
    space: 'Space',
    tab: 'Tab',
    enter: 'Enter',
    return: 'Enter',
    esc: 'Escape',
    escape: 'Escape',
    up: 'ArrowUp',
    down: 'ArrowDown',
    left: 'ArrowLeft',
    right: 'ArrowRight',
    backspace: 'Backspace',
    delete: 'Delete',
    slash: 'Slash',
    backslash: 'Backslash',
    minus: 'Minus',
    plus: 'Equal',
    equal: 'Equal',
    comma: 'Comma',
    period: 'Period',
    quote: 'Quote',
    semicolon: 'Semicolon',
    bracketleft: 'BracketLeft',
    bracketright: 'BracketRight',
    backquote: 'Backquote',
  };
  const lower = normalized.toLowerCase();
  if (exactMap[lower]) return exactMap[lower];
  if (
    /^key[a-z]$/i.test(normalized) ||
    /^digit[0-9]$/i.test(normalized) ||
    /^arrow(?:up|down|left|right)$/i.test(normalized) ||
    /^f\d{1,2}$/i.test(normalized)
  ) {
    const keyMatch = normalized.match(/^key([a-z])$/i);
    if (keyMatch) {
      return `Key${keyMatch[1]!.toUpperCase()}`;
    }

    const digitMatch = normalized.match(/^digit([0-9])$/i);
    if (digitMatch) {
      return `Digit${digitMatch[1]}`;
    }

    const arrowMatch = normalized.match(/^arrow(up|down|left|right)$/i);
    if (arrowMatch) {
      const direction = arrowMatch[1]!;
      return `Arrow${direction[0]!.toUpperCase()}${direction.slice(1).toLowerCase()}`;
    }

    const functionKeyMatch = normalized.match(/^f(\d{1,2})$/i);
    if (functionKeyMatch) {
      return `F${functionKeyMatch[1]}`;
    }
  }
  return null;
}

function parseAccelerator(
  accelerator: string,
  platform: PlatformKeyModel,
): { key: SessionKeySpec | null; message?: string } {
  const normalized = accelerator.replace(/\s+/g, '').replace(/cmdorctrl/gi, 'CommandOrControl');
  if (!normalized) {
    return { key: null, message: 'Empty accelerator is not supported.' };
  }

  const parts = normalized.split('+').filter(Boolean);
  const keyToken = parts.pop();
  if (!keyToken) {
    return { key: null, message: 'Missing accelerator key token.' };
  }

  const modifiers: SessionKeyModifier[] = [];
  for (const modifier of parts) {
    const lower = modifier.toLowerCase();
    if (lower === 'ctrl' || lower === 'control') {
      modifiers.push('ctrl');
      continue;
    }
    if (lower === 'alt' || lower === 'option') {
      modifiers.push('alt');
      continue;
    }
    if (lower === 'shift') {
      modifiers.push('shift');
      continue;
    }
    if (lower === 'meta' || lower === 'super' || lower === 'command' || lower === 'cmd') {
      modifiers.push('meta');
      continue;
    }
    if (lower === 'commandorcontrol') {
      modifiers.push(platform === 'darwin' ? 'meta' : 'ctrl');
      continue;
    }
    return {
      key: null,
      message: `Unsupported accelerator modifier: ${modifier}`,
    };
  }

  const code = normalizeCodeToken(keyToken);
  if (!code) {
    return {
      key: null,
      message: `Unsupported accelerator key token: ${keyToken}`,
    };
  }

  return {
    key: {
      code,
      modifiers: normalizeModifiers(modifiers),
    },
  };
}

function parseDomKeyString(
  key: string,
  platform: PlatformKeyModel,
): { key: SessionKeySpec | null; message?: string } {
  const parts = key
    .split('+')
    .map((part) => part.trim())
    .filter(Boolean);
  const keyToken = parts.pop();
  if (!keyToken) {
    return { key: null, message: 'Missing keybinding key token.' };
  }

  const modifiers: SessionKeyModifier[] = [];
  for (const modifier of parts) {
    const lower = modifier.toLowerCase();
    if (lower === 'ctrl' || lower === 'control') {
      modifiers.push('ctrl');
      continue;
    }
    if (lower === 'alt' || lower === 'option') {
      modifiers.push('alt');
      continue;
    }
    if (lower === 'shift') {
      modifiers.push('shift');
      continue;
    }
    if (
      lower === 'meta' ||
      lower === 'super' ||
      lower === 'command' ||
      lower === 'cmd' ||
      lower === 'commandorcontrol'
    ) {
      modifiers.push(
        lower === 'commandorcontrol' ? (platform === 'darwin' ? 'meta' : 'ctrl') : 'meta',
      );
      continue;
    }
    return {
      key: null,
      message: `Unsupported keybinding modifier: ${modifier}`,
    };
  }

  const code = normalizeCodeToken(keyToken, { allowMouseButtons: true });
  if (!code) {
    return {
      key: null,
      message: `Unsupported keybinding token: ${keyToken}`,
    };
  }

  return {
    key: {
      code,
      modifiers: normalizeModifiers(modifiers),
    },
  };
}

export function getSessionKeySpecSignature(key: SessionKeySpec): string {
  return [...key.modifiers, key.code].join('+');
}

function resolveCommandBinding(
  binding: Keybinding,
):
  | Omit<CompiledMpvCommandBinding, 'key' | 'sourcePath' | 'originalKey'>
  | Omit<CompiledSessionActionBinding, 'key' | 'sourcePath' | 'originalKey'>
  | null {
  const command = binding.command;
  if (!Array.isArray(command) || command.length === 0 || !command.every(isValidCommandEntry)) {
    return null;
  }

  const first = command[0];
  if (typeof first !== 'string') {
    return null;
  }

  if (first === SPECIAL_COMMANDS.SUBSYNC_TRIGGER) {
    if (command.length !== 1) return null;
    return { actionType: 'session-action', actionId: 'triggerSubsync' };
  }
  if (first === SPECIAL_COMMANDS.RUNTIME_OPTIONS_OPEN) {
    if (command.length !== 1) return null;
    return { actionType: 'session-action', actionId: 'openRuntimeOptions' };
  }
  if (first === SPECIAL_COMMANDS.JIMAKU_OPEN) {
    if (command.length !== 1) return null;
    return { actionType: 'session-action', actionId: 'openJimaku' };
  }
  if (first === SPECIAL_COMMANDS.TSUKIHIME_OPEN) {
    if (command.length !== 1) return null;
    return { actionType: 'session-action', actionId: 'openTsukihime' };
  }
  if (first === SPECIAL_COMMANDS.YOUTUBE_PICKER_OPEN) {
    if (command.length !== 1) return null;
    return { actionType: 'session-action', actionId: 'openYoutubePicker' };
  }
  if (first === SPECIAL_COMMANDS.PLAYLIST_BROWSER_OPEN) {
    if (command.length !== 1) return null;
    return { actionType: 'session-action', actionId: 'openPlaylistBrowser' };
  }
  if (first === SPECIAL_COMMANDS.REPLAY_SUBTITLE) {
    if (command.length !== 1) return null;
    return { actionType: 'session-action', actionId: 'replayCurrentSubtitle' };
  }
  if (first === SPECIAL_COMMANDS.PLAY_NEXT_SUBTITLE) {
    if (command.length !== 1) return null;
    return { actionType: 'session-action', actionId: 'playNextSubtitle' };
  }
  if (first.startsWith(SPECIAL_COMMANDS.RUNTIME_OPTION_CYCLE_PREFIX)) {
    if (command.length !== 1) {
      return null;
    }
    const parts = first.split(':');
    if (parts.length !== 3) {
      return null;
    }
    const [, runtimeOptionId, rawDirection] = parts;
    if (!runtimeOptionId || (rawDirection !== 'prev' && rawDirection !== 'next')) {
      return null;
    }
    return {
      actionType: 'session-action',
      actionId: 'cycleRuntimeOption',
      payload: {
        runtimeOptionId,
        direction: rawDirection === 'prev' ? -1 : 1,
      },
    };
  }

  return {
    actionType: 'mpv-command',
    command,
  };
}

function getBindingFingerprint(binding: CompiledSessionBinding): string {
  if (binding.actionType === 'mpv-command') {
    return `mpv:${JSON.stringify(binding.command)}`;
  }
  return `session:${binding.actionId}:${JSON.stringify(binding.payload ?? null)}`;
}

function buildSessionActionCliArgs(binding: CompiledSessionActionBinding): string[] {
  const request =
    binding.payload === undefined
      ? { actionId: binding.actionId }
      : { actionId: binding.actionId, payload: binding.payload };
  return ['--session-action', JSON.stringify(request)];
}

function toPluginSessionBinding(binding: CompiledSessionBinding): PluginSessionBinding {
  if (binding.actionType !== 'session-action') {
    return binding;
  }

  if (
    (binding.actionId === 'copySubtitleMultiple' || binding.actionId === 'mineSentenceMultiple') &&
    binding.payload?.count === undefined
  ) {
    return binding;
  }

  return { ...binding, cliArgs: buildSessionActionCliArgs(binding) };
}

export function compileSessionBindings(input: CompileSessionBindingsInput): {
  bindings: CompiledSessionBinding[];
  warnings: SessionBindingWarning[];
} {
  const warnings: SessionBindingWarning[] = [];
  const candidates = new Map<string, DraftBinding[]>();
  const legacyToggleVisibleOverlayGlobal = (
    input.rawConfig?.shortcuts as Record<string, unknown> | undefined
  )?.toggleVisibleOverlayGlobal;
  const statsToggleKey = input.statsToggleKey ?? input.rawConfig?.stats?.toggleKey ?? null;
  const statsMarkWatchedKey =
    input.statsMarkWatchedKey ?? input.rawConfig?.stats?.markWatchedKey ?? null;

  if (legacyToggleVisibleOverlayGlobal !== undefined) {
    warnings.push({
      kind: 'deprecated-config',
      path: 'shortcuts.toggleVisibleOverlayGlobal',
      value: legacyToggleVisibleOverlayGlobal,
      message: 'Rename shortcuts.toggleVisibleOverlayGlobal to shortcuts.toggleVisibleOverlay.',
    });
  }

  for (const shortcut of SESSION_SHORTCUT_ACTIONS) {
    const accelerator = input.shortcuts[shortcut.key];
    if (!accelerator) continue;
    const parsed = parseAccelerator(accelerator, input.platform);
    if (!parsed.key) {
      warnings.push({
        kind: 'unsupported',
        path: `shortcuts.${shortcut.key}`,
        value: accelerator,
        message: parsed.message ?? 'Unsupported accelerator syntax.',
      });
      continue;
    }
    const binding: CompiledSessionActionBinding = {
      sourcePath: `shortcuts.${shortcut.key}`,
      originalKey: accelerator,
      key: parsed.key,
      actionType: 'session-action',
      actionId: shortcut.actionId,
    };
    const signature = getSessionKeySpecSignature(parsed.key);
    const draft = candidates.get(signature) ?? [];
    draft.push({
      binding,
      actionFingerprint: getBindingFingerprint(binding),
    });
    candidates.set(signature, draft);
  }

  if (statsToggleKey) {
    const parsed = parseDomKeyString(statsToggleKey, input.platform);
    if (!parsed.key) {
      warnings.push({
        kind: 'unsupported',
        path: 'stats.toggleKey',
        value: statsToggleKey,
        message: parsed.message ?? 'Unsupported stats toggle key syntax.',
      });
    } else {
      const binding: CompiledSessionActionBinding = {
        sourcePath: 'stats.toggleKey',
        originalKey: statsToggleKey,
        key: parsed.key,
        actionType: 'session-action',
        actionId: 'toggleStatsOverlay',
      };
      const signature = getSessionKeySpecSignature(parsed.key);
      const draft = candidates.get(signature) ?? [];
      draft.push({
        binding,
        actionFingerprint: getBindingFingerprint(binding),
      });
      candidates.set(signature, draft);
    }
  }

  if (statsMarkWatchedKey) {
    const parsed = parseDomKeyString(statsMarkWatchedKey, input.platform);
    if (!parsed.key) {
      warnings.push({
        kind: 'unsupported',
        path: 'stats.markWatchedKey',
        value: statsMarkWatchedKey,
        message: parsed.message ?? 'Unsupported stats mark-watched key syntax.',
      });
    } else {
      const binding: CompiledSessionActionBinding = {
        sourcePath: 'stats.markWatchedKey',
        originalKey: statsMarkWatchedKey,
        key: parsed.key,
        actionType: 'session-action',
        actionId: 'markWatched',
      };
      const signature = getSessionKeySpecSignature(parsed.key);
      const draft = candidates.get(signature) ?? [];
      draft.push({
        binding,
        actionFingerprint: getBindingFingerprint(binding),
      });
      candidates.set(signature, draft);
    }
  }

  input.keybindings.forEach((binding, index) => {
    if (!binding.command) return;
    const parsed = parseDomKeyString(binding.key, input.platform);
    if (!parsed.key) {
      warnings.push({
        kind: 'unsupported',
        path: `keybindings[${index}].key`,
        value: binding.key,
        message: parsed.message ?? 'Unsupported keybinding syntax.',
      });
      return;
    }
    const resolved = resolveCommandBinding(binding);
    if (!resolved) {
      warnings.push({
        kind: 'unsupported',
        path: `keybindings[${index}].command`,
        value: binding.command,
        message: 'Unsupported keybinding command syntax.',
      });
      return;
    }
    const compiled: CompiledSessionBinding = {
      sourcePath: `keybindings[${index}].key`,
      originalKey: binding.key,
      key: parsed.key,
      ...resolved,
    };
    const signature = getSessionKeySpecSignature(parsed.key);
    const draft = candidates.get(signature) ?? [];
    draft.push({
      binding: compiled,
      actionFingerprint: getBindingFingerprint(compiled),
    });
    candidates.set(signature, draft);
  });

  const bindings: CompiledSessionBinding[] = [];
  for (const [signature, draftBindings] of candidates.entries()) {
    const uniqueFingerprints = new Set(draftBindings.map((entry) => entry.actionFingerprint));
    if (uniqueFingerprints.size > 1) {
      warnings.push({
        kind: 'conflict',
        path: draftBindings[0]!.binding.sourcePath,
        value: signature,
        conflictingPaths: draftBindings.map((entry) => entry.binding.sourcePath),
        message: `Conflicting session bindings compile to ${signature}; SubMiner will bind neither action.`,
      });
      continue;
    }
    bindings.push(draftBindings[0]!.binding);
  }

  bindings.sort((left, right) => left.sourcePath.localeCompare(right.sourcePath));
  return { bindings, warnings };
}

export function buildPluginSessionBindingsArtifact(input: {
  bindings: CompiledSessionBinding[];
  warnings: SessionBindingWarning[];
  numericSelectionTimeoutMs: number;
  now?: Date;
}): PluginSessionBindingsArtifact {
  return {
    version: 1,
    generatedAt: (input.now ?? new Date()).toISOString(),
    numericSelectionTimeoutMs: input.numericSelectionTimeoutMs,
    bindings: input.bindings.map(toPluginSessionBinding),
    warnings: input.warnings,
  };
}

File: plugin/subminer/session_bindings.lua
local M = {}

local unpack_fn = table.unpack or unpack

local KEY_NAME_MAP = {
	Space = "SPACE",
	Tab = "TAB",
	Enter = "ENTER",
	Escape = "ESC",
	Backspace = "BS",
	Delete = "DEL",
	ArrowUp = "UP",
	ArrowDown = "DOWN",
	ArrowLeft = "LEFT",
	ArrowRight = "RIGHT",
	Slash = "/",
	Backslash = "\\",
	Minus = "-",
	Equal = "=",
	Comma = ",",
	Period = ".",
	Quote = "'",
	Semicolon = ";",
	BracketLeft = "[",
	BracketRight = "]",
	Backquote = "`",
	MBTN_LEFT = "MBTN_LEFT",
	MBTN_MID = "MBTN_MID",
	MBTN_RIGHT = "MBTN_RIGHT",
	MBTN_BACK = "MBTN_BACK",
	MBTN_FORWARD = "MBTN_FORWARD",
}

local MODIFIER_MAP = {
	ctrl = "Ctrl",
	alt = "Alt",
	shift = "Shift",
	meta = "Meta",
}

local SHIFTED_KEY_NAME_MAP = {
	Digit1 = "!",
	Digit2 = "@",
	Digit3 = "SHARP",
	Digit4 = "$",
	Digit5 = "%",
	Digit6 = "^",
	Digit7 = "&",
	Digit8 = "*",
	Digit9 = "(",
	Digit0 = ")",
	Minus = "_",
	Equal = "+",
	BracketLeft = "{",
	BracketRight = "}",
	Backslash = "|",
	Semicolon = ":",
	Quote = '"',
	Comma = "<",
	Period = ">",
	Slash = "?",
	Backquote = "~",
}

function M.create(ctx)
	local mp = ctx.mp
	local utils = ctx.utils
	local state = ctx.state
	local process = ctx.process
	local environment = ctx.environment
	local subminer_log = ctx.log.subminer_log
	local show_osd = ctx.log.show_osd

	local function read_file(path)
		local handle = io.open(path, "r")
		if not handle then
			return nil
		end
		local content = handle:read("*a")
		handle:close()
		return content
	end

	local function remove_binding_names(names)
		for _, name in ipairs(names) do
			mp.remove_key_binding(name)
		end
		for index = #names, 1, -1 do
			names[index] = nil
		end
	end

	local function key_code_to_mpv_name(code)
		if KEY_NAME_MAP[code] then
			return KEY_NAME_MAP[code]
		end

		local letter = code:match("^Key([A-Z])$")
		if letter then
			return string.lower(letter)
		end

		local digit = code:match("^Digit([0-9])$")
		if digit then
			return digit
		end

		local function_key = code:match("^(F%d+)$")
		if function_key then
			return function_key
		end

		return nil
	end

	local function contains_value(values, target)
		for _, value in ipairs(values) do
			if value == target then
				return true
			end
		end
		return false
	end

	local function append_unique(values, value)
		if not contains_value(values, value) then
			values[#values + 1] = value
		end
	end

	local function key_spec_to_mpv_bindings(key)
		if type(key) ~= "table" then
			return nil
		end

		if type(key.code) ~= "string" then
			return nil
		end
		if type(key.modifiers) ~= "table" then
			return nil
		end

		local shifted_letter = key.code:match("^Key([A-Z])$")
		local has_shift = false
		for _, modifier in ipairs(key.modifiers) do
			if modifier == "shift" then
				has_shift = true
				break
			end
		end

		local key_name = key_code_to_mpv_name(key.code)
		if shifted_letter and has_shift then
			key_name = shifted_letter
		end
		if not key_name then
			return nil
		end

		local parts = {}
		for _, modifier in ipairs(key.modifiers) do
			if not (modifier == "shift" and shifted_letter) then
				local mapped = MODIFIER_MAP[modifier]
				if mapped then
					parts[#parts + 1] = mapped
				end
			end
		end
		parts[#parts + 1] = key_name
		local bindings = { table.concat(parts, "+") }

		local shifted_key_name = SHIFTED_KEY_NAME_MAP[key.code]
		if has_shift and shifted_key_name then
			local shifted_parts = {}
			for _, modifier in ipairs(key.modifiers) do
				if modifier ~= "shift" then
					local mapped = MODIFIER_MAP[modifier]
					if mapped then
						shifted_parts[#shifted_parts + 1] = mapped
					end
				end
			end
			shifted_parts[#shifted_parts + 1] = shifted_key_name
			append_unique(bindings, table.concat(shifted_parts, "+"))
		end

		return bindings
	end

	local function normalize_cli_args(cli_args)
		if type(cli_args) ~= "table" then
			return nil
		end

		local normalized = {}
		for _, arg in ipairs(cli_args) do
			if type(arg) ~= "string" and type(arg) ~= "number" then
				return nil
			end
			local value = tostring(arg)
			if value == "" then
				return nil
			end
			normalized[#normalized + 1] = value
		end

		if #normalized == 0 then
			return nil
		end
		return normalized
	end

	local function build_cli_args(action_id, payload, artifact_cli_args)
		local cli_args = normalize_cli_args(artifact_cli_args)
		if cli_args then
			return cli_args
		end

		if action_id == "toggleVisibleOverlay" then
			return { "--toggle-visible-overlay" }
		elseif action_id == "toggleStatsOverlay" then
			return { "--toggle-stats-overlay" }
		elseif action_id == "copySubtitle" then
			return { "--copy-subtitle" }
		elseif action_id == "copySubtitleMultiple" then
			if payload and payload.count then
				return { "--copy-subtitle-count", tostring(payload.count) }
			end
			return { "--copy-subtitle-multiple" }
		elseif action_id == "updateLastCardFromClipboard" then
			return { "--update-last-card-from-clipboard" }
		elseif action_id == "triggerFieldGrouping" then
			return { "--trigger-field-grouping" }
		elseif action_id == "triggerSubsync" then
			return { "--session-action", '{"actionId":"triggerSubsync"}' }
		elseif action_id == "mineSentence" then
			return { "--mine-sentence" }
		elseif action_id == "mineSentenceMultiple" then
			if payload and payload.count then
				return { "--mine-sentence-count", tostring(payload.count) }
			end
			return { "--mine-sentence-multiple" }
		elseif action_id == "toggleSecondarySub" then
			return { "--toggle-secondary-sub" }
		elseif action_id == "toggleSubtitleSidebar" then
			return { "--toggle-subtitle-sidebar" }
		elseif action_id == "toggleNotificationHistory" then
			return { "--session-action", '{"actionId":"toggleNotificationHistory"}' }
		elseif action_id == "markAudioCard" then
			return { "--mark-audio-card" }
		elseif action_id == "markWatched" then
			return { "--mark-watched" }
		elseif action_id == "openRuntimeOptions" then
			return { "--session-action", '{"actionId":"openRuntimeOptions"}' }
		elseif action_id == "openJimaku" then
			return { "--open-jimaku" }
		elseif action_id == "openTsukihime" or action_id == "openAnimetosho" then
			return { "--open-tsukihime" }
		elseif action_id == "openYoutubePicker" then
			return { "--open-youtube-picker" }
		elseif action_id == "openSessionHelp" then
			return { "--session-action", '{"actionId":"openSessionHelp"}' }
		elseif action_id == "openCharacterDictionaryManager" then
			return { "--session-action", '{"actionId":"openCharacterDictionaryManager"}' }
		elseif action_id == "openControllerSelect" then
			return { "--open-controller-select" }
		elseif action_id == "openControllerDebug" then
			return { "--open-controller-debug" }
		elseif action_id == "openPlaylistBrowser" then
			return { "--open-playlist-browser" }
		elseif action_id == "replayCurrentSubtitle" then
			return { "--replay-current-subtitle" }
		elseif action_id == "playNextSubtitle" then
			return { "--play-next-subtitle" }
		elseif action_id == "cycleRuntimeOption" then
			local runtime_option_id = payload and payload.runtimeOptionId or nil
			if type(runtime_option_id) ~= "string" or runtime_option_id == "" then
				return nil
			end
			local direction = payload and payload.direction == -1 and "prev" or "next"
			return { "--cycle-runtime-option", runtime_option_id .. ":" .. direction }
		end

		return nil
	end

	local function invoke_cli_action(action_id, payload, artifact_cli_args)
		if not process.check_binary_available() then
			show_osd("Error: binary not found")
			return
		end

		local cli_args = build_cli_args(action_id, payload, artifact_cli_args)
		if not cli_args then
			subminer_log("warn", "session-bindings", "No CLI mapping for action: " .. tostring(action_id))
			return
		end

		local args = { state.binary_path }
		for _, arg in ipairs(cli_args) do
			args[#args + 1] = arg
		end
		local runner = process.run_binary_command_async
		if type(runner) ~= "function" then
			runner = function(binary_args, callback)
				mp.command_native_async({
					name = "subprocess",
					args = binary_args,
					playback_only = false,
					capture_stdout = true,
					capture_stderr = true,
				}, function(success, result, error)
					local ok = success and (result == nil or result.status == 0)
					if callback then
						callback(ok, result, error)
					end
				end)
			end
		end
		runner(args, function(ok, result, error)
			if ok then
				return
			end
			local reason = error or (result and result.stderr) or "unknown error"
			subminer_log("warn", "session-bindings", "Session action failed: " .. tostring(reason))
			show_osd("Session action failed")
		end)
	end

	local function execute_mpv_command(command)
		if type(command) ~= "table" or command[1] == nil then
			return
		end
		mp.commandv(unpack_fn(command))
	end

	local function handle_binding(binding)
		if binding.actionType == "mpv-command" then
			execute_mpv_command(binding.command)
			return
		end

		if binding.actionId == "toggleVisibleOverlay" and type(process.toggle_overlay) == "function" then
			process.toggle_overlay()
			return
		end

		invoke_cli_action(binding.actionId, binding.payload, binding.cliArgs)
	end

	local function is_supported_binding(binding)
		if binding.actionType == "mpv-command" then
			return type(binding.command) == "table" and binding.command[1] ~= nil
		end
		if binding.actionType == "session-action" then
			return build_cli_args(binding.actionId, binding.payload, binding.cliArgs) ~= nil
		end
		return false
	end

	local function load_artifact()
		local artifact_path = environment.resolve_session_bindings_artifact_path()
		local raw = read_file(artifact_path)
		if not raw or raw == "" then
			return nil, "Missing session binding artifact: " .. tostring(artifact_path)
		end

		local parsed, parse_error = utils.parse_json(raw)
		if not parsed then
			return nil, "Failed to parse session binding artifact: " .. tostring(parse_error)
		end
		if type(parsed) ~= "table" or type(parsed.bindings) ~= "table" then
			return nil, "Invalid session binding artifact"
		end

		return parsed, nil
	end

	local function clear_bindings()
		remove_binding_names(state.session_binding_names)
	end

	local function register_bindings()
		local artifact, load_error = load_artifact()
		if not artifact then
			subminer_log("warn", "session-bindings", load_error)
			return false
		end

		local previous_binding_names = state.session_binding_names
		local next_binding_names = {}
		state.session_binding_generation = (state.session_binding_generation or 0) + 1
		local generation = state.session_binding_generation

		for index, binding in ipairs(artifact.bindings) do
			if not is_supported_binding(binding) then
				subminer_log(
					"warn",
					"session-bindings",
					"Skipped unsupported session binding from artifact"
				)
			else
				local key_names = key_spec_to_mpv_bindings(binding.key)
				if key_names then
					for key_index, key_name in ipairs(key_names) do
						local name = "subminer-session-binding-"
							.. tostring(generation)
							.. "-"
							.. tostring(index)
							.. "-"
							.. tostring(key_index)
						next_binding_names[#next_binding_names + 1] = name
						mp.add_forced_key_binding(key_name, name, function()
							handle_binding(binding)
						end)
					end
				else
					subminer_log(
						"warn",
						"session-bindings",
						"Skipped unsupported key code from artifact: " .. tostring(binding.key and binding.key.code or "unknown")
					)
				end
			end
		end

		remove_binding_names(previous_binding_names)
		state.session_binding_names = next_binding_names

		subminer_log(
			"info",
			"session-bindings",
			"Registered " .. tostring(#next_binding_names) .. " shared session bindings"
		)
		return true
	end

	local function reload_bindings()
		return register_bindings()
	end

	return {
		register_bindings = register_bindings,
		reload_bindings = reload_bindings,
		clear_bindings = clear_bindings,
	}
end

return M
